gh-75459: Doc: C API: Improve object life cycle documentation (#125962)
python/cpython: Doc/c-api/typeobj.rst

.. c:member:: destructor PyTypeObject.tp_dealloc

   A pointer to the instance destructor function.  The function signature is::

      void tp_dealloc(PyObject *self);

   The destructor function should free all references which the instance owns,
   free all memory buffers owned by the instance (using the freeing function
   corresponding to the allocation function used to allocate the buffer), and
   call the type's :c:member:`~PyTypeObject.tp_free` function to free the
   object itself.

   If the type is not subtypable (the :c:macro:`Py_TPFLAGS_BASETYPE` flag is
   not set), it is permissible to call the object deallocator directly instead
   of via :c:member:`~PyTypeObject.tp_free`.  The called deallocator should be
   the one that pairs with the allocator used to allocate the instance; this is
   normally :c:func:`PyObject_GC_Del` if the type supports garbage collection
   (the :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set) or :c:func:`PyObject_Free`
   otherwise.

   No guarantees are made about when an object is destroyed, except:

>    * Python will destroy an object after the final reference to the object is

Reviewer: ```suggestion
   * Python will destroy an object some time after the final reference to the object is
```

.. c:member:: destructor PyTypeObject.tp_dealloc

   A pointer to the instance destructor function.  The function signature is::

      void tp_dealloc(PyObject *self);

>    The destructor function should free all references which the instance owns,

Reviewer 2: I don't want to mix up freeing memory and decrementing a reference count
```suggestion
   The destructor function should release all references (i.e., call :c:func:`Py_DECREF`) which the instance owns,
```

   free all memory buffers owned by the instance (using the freeing function

>    corresponding to the allocation function used to allocate the buffer), and

Reviewer 2: Freeing memory using the proper allocator seems implied to me:
```suggestion
   free all memory buffers owned by the instance, and
```

   call the type's :c:member:`~PyTypeObject.tp_free` function to free the
   object itself.

   If the type is not subtypable (the :c:macro:`Py_TPFLAGS_BASETYPE` flag is
   not set), it is permissible to call the object deallocator directly instead
   of via :c:member:`~PyTypeObject.tp_free`.  The called deallocator should be
   the one that pairs with the allocator used to allocate the instance; this is
   normally :c:func:`PyObject_GC_Del` if the type supports garbage collection
   (the :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set) or :c:func:`PyObject_Free`

>    otherwise.

Reviewer 2: I'd like to remove this section. Even if the object can't be subclassed, we shouldn't encourage users to manually call object allocator functions. It makes for more maintenance work if they decide it can be a basetype in the future (or change it from non-GC to GC or something like that), and also limits our ability to change things based on our assumptions of what users are doing with object allocation.

   No guarantees are made about when an object is destroyed, except:

   * Python will destroy an object immediately or some time after the final
     reference to the object is deleted, unless its finalizer
     (:c:member:`~PyTypeObject.tp_finalize`) subsequently resurrects the
     object.
   * An object will not be destroyed while it is being automatically finalized
     (:c:member:`~PyTypeObject.tp_finalize`) or automatically cleared
     (:c:member:`~PyTypeObject.tp_clear`).

   .. warning::

      The :c:member:`~PyTypeObject.tp_dealloc` function can be called from any
      thread, although the :term:`GIL` will be held unless disabled (see

>       :ref:`whatsnew313-free-threaded-cpython`).

Reviewer 2: "GIL" in the C API is a bit of a loose term. Really, the important part is that the calling thread has an active thread state, and that will be true even if the GIL is disabled.

```suggestion
      thread, although the :term:`GIL` will be held.
```

>    :term:`CPython` currently destroys an object immediately from

Reviewer 2: This is all CPython-specific, no need to explicitly use the term.
```suggestion
   Python currently destroys an object immediately from
```

   :c:func:`Py_DECREF` when the new reference count is zero, but this may
   change in a future version.

   It is recommended to call :c:func:`PyObject_CallFinalizerFromDealloc` at the
   beginning of :c:member:`!tp_dealloc` to guarantee that the object is always
   finalized before destruction.

   If the type supports garbage collection (the :c:macro:`Py_TPFLAGS_HAVE_GC`
   flag is set), the destructor should call :c:func:`PyObject_GC_UnTrack`
   before clearing any member fields.

   It is permissible to call :c:member:`~PyTypeObject.tp_clear` from
   :c:member:`!tp_dealloc` to reduce code duplication and to guarantee that the
   object is always cleared before destruction.  Beware that

>    :c:member:`!tp_clear` might have already been called.

Reviewer 2: It might be worth noting that `Py_CLEAR` has a `NULL` check anyway, so it's normally safe to call it multiple times.

   If the type is heap allocated (:c:macro:`Py_TPFLAGS_HEAPTYPE`), the
   deallocator should release the owned reference to its type object (via

>    :c:func:`Py_DECREF`) after calling the type deallocator.

Reviewer 2: I think an example would be nice.

   .. warning::

      The :c:member:`~PyTypeObject.tp_dealloc` function can be called from any
      thread, although the :term:`GIL` will be held.

   Python currently destroys an object immediately from :c:func:`Py_DECREF`
   when the new reference count is zero, but this may change in a future
   version.

   It is recommended to call :c:func:`PyObject_CallFinalizerFromDealloc` at the
   beginning of :c:member:`!tp_dealloc` to guarantee that the object is always
   finalized before destruction.

   If the type supports garbage collection (the :c:macro:`Py_TPFLAGS_HAVE_GC`
   flag is set), the destructor should call :c:func:`PyObject_GC_UnTrack`
   before clearing any member fields.

   It is permissible to call :c:member:`~PyTypeObject.tp_clear` from
   :c:member:`!tp_dealloc` to reduce code duplication and to guarantee that the
   object is always cleared before destruction.  Beware that
   :c:member:`!tp_clear` might have already been called.

   If the type is heap allocated (:c:macro:`Py_TPFLAGS_HEAPTYPE`), the
   deallocator should release the owned reference to its type object (via
   :c:func:`Py_DECREF`) after calling the type deallocator.  See the example

>    code below.

Reviewer 3: Doctest failures: we still need the `.. code-block:: c`, or because we already have C as the default for the document (set at the top), add `::` at the end of this paragraph.

.. c:member:: destructor PyTypeObject.tp_dealloc

   A pointer to the instance destructor function.  The function signature is::

      void tp_dealloc(PyObject *self);

   The destructor function should remove all references which the instance owns
   (e.g., call :c:func:`Py_CLEAR`), free all memory buffers owned by the
   instance, and call the type's :c:member:`~PyTypeObject.tp_free` function to
   free the object itself.

   No guarantees are made about when an object is destroyed, except:

   * Python will destroy an object immediately or some time after the final
     reference to the object is deleted, unless its finalizer
     (:c:member:`~PyTypeObject.tp_finalize`) subsequently resurrects the
     object.
   * An object will not be destroyed while it is being automatically finalized
     (:c:member:`~PyTypeObject.tp_finalize`) or automatically cleared
     (:c:member:`~PyTypeObject.tp_clear`).

   .. warning::

      The :c:member:`~PyTypeObject.tp_dealloc` function can be called from any

>       thread, although the :term:`GIL` will be held.

Reviewer: This does not need to be a warning.
There's a paragraph below that says this in a bit more detail (and using the  “attached thread state” terminology).
```suggestion
```

>    Python currently destroys an object immediately from :c:func:`Py_DECREF`

Reviewer: ```suggestion
   CPython currently destroys an object immediately from :c:func:`Py_DECREF`
```

   with :c:func:`PyErr_WriteUnraisable`).

   Example::

      static void
      foo_dealloc(PyObject *self)
      {
          PyObject *exc = PyErr_GetRaisedException();

          if (PyObject_CallFinalizerFromDealloc(self) < 0) {
              // self was resurrected.
              goto done;
          }

          PyTypeObject *tp = Py_TYPE(self);

          if (tp->tp_flags & Py_TPFLAGS_HAVE_GC) {
              PyObject_GC_UnTrack(self);
          }

          // Optional, but convenient to avoid code duplication.
          if (tp->tp_clear && tp->tp_clear(self) < 0) {
              PyErr_WriteUnraisable(self);
          }

          // Any additional destruction goes here.

          tp->tp_free(self);
          self = NULL;  // In case PyErr_WriteUnraisable() is called below.

          if (tp->tp_flags & Py_TPFLAGS_HEAPTYPE) {
              Py_CLEAR(tp);
          }

      done:
          // Optional, if something was called that might have raised an
          // exception.
          if (PyErr_Occurred()) {
              PyErr_WriteUnraisable(self);
          }
          PyErr_SetRaisedException(exc);
      }

>    In a garbage collected Python, :c:member:`!tp_dealloc` may be called from

Reviewer: This should be true anywhere -- the thread that created the object might no longer exist.
```suggestion
   :c:member:`!tp_dealloc` may be called from
```

   any Python thread, not just the thread which created the object (if the
   object becomes part of a refcount cycle, that cycle might be collected by
   a garbage collection on any thread).  This is not a problem for Python
   API calls, since the thread on which :c:member:`!tp_dealloc` is called
   with an :term:`attached thread state`.  However, if the object being

>    destroyed in turn destroys objects from some other C or C++ library, care

Reviewer: Or Rust or Erlang.
```suggestion
   destroyed in turn destroys objects from some other C library, care
```
(for pedants: that's as in "C API", not "C language")

   .. c:macro:: Py_TPFLAGS_HAVE_GC

      This bit is set when the object supports garbage collection.  If this bit
      is set, memory for new instances (see :c:member:`~PyTypeObject.tp_alloc`)
      must be allocated using :c:macro:`PyObject_GC_New` or
      :c:func:`PyType_GenericAlloc` and deallocated (see
      :c:member:`~PyTypeObject.tp_free`) using :c:func:`PyObject_GC_Del`.  More
      information in section :ref:`supporting-cycle-detection`.  This bit also
      implies that the GC-related fields :c:member:`~PyTypeObject.tp_traverse`

>       and :c:member:`~PyTypeObject.tp_clear` are present in the type object.

Reviewer: The field existence notes are no longer needed. (Python 2 attempted to be ABI-compatible between feature releases, and used these feature flags for that.)
```suggestion
      information in section :ref:`supporting-cycle-detection`.
```

   heap-allocated superclass).
   If they do not, the type object may not be garbage-collected.

   .. warning::

      The :c:member:`~PyTypeObject.tp_traverse` function can be called from any

>       thread, although the :term:`GIL` will be held unless disabled (see
>       :ref:`whatsnew313-free-threaded-cpython`).

Reviewer 2: Per my other comments:
```suggestion
      thread.
```

      The :c:member:`~PyTypeObject.tp_clear` function can be called from any
      thread, although the :term:`GIL` will be held.

   .. warning::

      An object is not guaranteed to be automatically cleared before its
      destructor (:c:member:`~PyTypeObject.tp_dealloc`) is called.

   This function differs from the destructor
   (:c:member:`~PyTypeObject.tp_dealloc`) in the following ways:

   * The purpose of clearing an object is to remove references to other objects
     that might participate in a reference cycle.  The purpose of the
     destructor, on the other hand, is a superset: it must release *all*
     resources it owns, including references to objects that cannot participate
     in a reference cycle (e.g., integers) as well as the object's own memory
     (by calling :c:member:`~PyTypeObject.tp_free`).
   * When :c:member:`!tp_clear` is called, other objects might still hold
     references to the object being cleared.  Because of this,
     :c:member:`!tp_clear` must not deallocate the object's own memory
     (:c:member:`~PyTypeObject.tp_free`).  The destructor, on the other hand,
     is only called when no (strong) references exist, and is thus able to

>      safely destroy the object itself.

Reviewer: In fact it *must* destroy the object (unless it was resurrected).

   heap-allocated superclass).
   If they do not, the type object may not be garbage-collected.

   .. warning::

      The :c:member:`~PyTypeObject.tp_traverse` function can be called from any

>       thread, although the :term:`GIL` will be held.

Reviewer: ```suggestion
   .. note::

      The :c:member:`~PyTypeObject.tp_traverse` function can be called from any
      thread.
```

.. c:member:: inquiry PyTypeObject.tp_clear

   An optional pointer to a clear function.  The signature is::

      int tp_clear(PyObject *);

   The purpose of this function is to break reference cycles that are causing a
   :term:`cyclic isolate` so that the objects can be safely destroyed.  A
   cleared object is a partially destroyed object; the object is not obligated
   to satisfy design invariants held during normal use.

   :c:member:`!tp_clear` does not need to delete references to objects that
   can't participate in reference cycles, such as Python strings or Python
   integers.  However, it may be convenient to clear all references, and write
   the type's :c:member:`~PyTypeObject.tp_dealloc` function to invoke
   :c:member:`!tp_clear` to avoid code duplication.  (Beware that
   :c:member:`!tp_clear` might have already been called, but :c:func:`Py_CLEAR`

>    is idempotent so extra checks are usually unnecessary.)

Reviewer: ```suggestion
   :c:member:`!tp_clear` might have already been called. Prefer calling
   idempotent functions like :c:func:`Py_CLEAR`.)
```

   Any non-trivial cleanup should be performed in
   :c:member:`~PyTypeObject.tp_finalize` instead of :c:member:`!tp_clear`.

>    .. warning::

Reviewer: ```suggestion
   .. note::
```

      If :c:member:`!tp_clear` fails to break a reference cycle then the
      objects in the :term:`cyclic isolate` may remain indefinitely
      uncollectable ("leak").  See :data:`gc.garbage`.

>    .. warning::

Reviewer: ```suggestion
   .. note::
```

      Referents (direct and indirect) might have already been cleared; they are
      not guaranteed to be in a consistent state.

>    .. warning::

Reviewer: ```suggestion
   .. note::
```

      The :c:member:`~PyTypeObject.tp_clear` function can be called from any

>       thread, although the :term:`GIL` will be held.

Reviewer: ```suggestion
      thread.
```

>    .. warning::

Reviewer: ```suggestion
   .. note::
```

   No guarantees are made about when, if, or how often Python automatically
   clears an object, except:

   * Python will not automatically clear an object if it is reachable, i.e.,
     there is a reference to it and it is not a member of a :term:`cyclic
     isolate`.
   * Python will not automatically clear an object if it has not been
     automatically finalized (see :c:member:`~PyTypeObject.tp_finalize`).  (If
     the finalizer resurrected the object, the object may or may not be
     automatically finalized again before it is cleared.)
   * If an object is a member of a :term:`cyclic isolate`, Python will not
     automatically clear it if any member of the cyclic isolate has not yet
     been automatically finalized (:c:member:`~PyTypeObject.tp_finalize`).
   * Python will not destroy an object until after any automatic calls to its
     :c:member:`!tp_clear` function have returned.  This ensures that the act
     of breaking a reference cycle does not invalidate the ``self`` pointer
     while :c:member:`!tp_clear` is still executing.
   * Python will not automatically call :c:member:`!tp_clear` multiple times
     concurrently.

>    Python currently only automatically clears objects as needed to break

Reviewer: As a rule tof thumb, things that may break in future versions are CPython-specific.
```suggestion
   CPython currently only automatically clears objects as needed to break
```

   **Inheritance:**

   Static subtypes inherit this slot, which will be
   :c:func:`PyType_GenericAlloc` if inherited from :class:`object`.

>    Dynamic subtypes (created by a class statement) do not inherit this slot.

Reviewer: In Python 3, you can now create dynamic types from C as well!
```suggestion
   :ref:`Heap subtypes <heap-types>` do not inherit this slot.
```

   **Default:**

>    For dynamic subtypes, this field is always set to

Reviewer: ```suggestion
   For heap subtypes, this field is always set to
```

      void tp_free(void *self);

   This function must free the memory allocated by
   :c:member:`~PyTypeObject.tp_alloc`.

   **Inheritance:**

   Static subtypes inherit this slot, which will be :c:func:`PyObject_Free` if
   inherited from :class:`object`.  Exception: If the type supports garbage
   collection (i.e., the :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set in
   :c:member:`~PyTypeObject.tp_flags`) and it would inherit
   :c:func:`PyObject_Free`, then this slot is not inherited but instead defaults
   to :c:func:`PyObject_GC_Del`.

>    Dynamic subtypes (created by a class statement) do not inherit this slot.

Reviewer: ```suggestion
   :ref:`Heap subtypes <heap-types>` do not inherit this slot.
```

   **Default:**

>    For dynamic subtypes, this slot defaults to a deallocator suitable to match

Reviewer: ```suggestion
   For :ref:`heap subtypes <heap-types>`, this slot defaults to a deallocator suitable to match
```

   consistent state.  See :pep:`442`.

   Before Python automatically finalizes an object, some of the object's direct
   or indirect referents might have themselves been automatically finalized.
   However, none of the referents will have been automatically cleared
   (:c:member:`~PyTypeObject.tp_clear`) yet.

   Other non-finalized objects might still be using a finalized object, so the
   finalizer must leave the object in a sane state (e.g., invariants are still
   met).

   .. warning::

      After Python automatically finalizes an object, Python might start
      automatically clearing (:c:member:`~PyTypeObject.tp_clear`) the object
      and its referents (direct and indirect).  Cleared objects are not
      guaranteed to be in a consistent state; a finalized object must be able
      to tolerate cleared referents.

   .. warning::

      An object is not guaranteed to be automatically finalized before its
      destructor (:c:member:`~PyTypeObject.tp_dealloc`) is called.  It is
      recommended to call :c:func:`PyObject_CallFinalizerFromDealloc` at the
      beginning of :c:member:`!tp_dealloc` to guarantee that the object is
      always finalized before destruction.

>    .. warning::
> 
>       The :c:member:`~PyTypeObject.tp_finalize` function can be called from any
>       thread, although the :term:`GIL` will be held unless disabled (see
>       :ref:`whatsnew313-free-threaded-cpython`).
> 

Reviewer 2: I think this has been repeated a few times already.
```suggestion
```

Author: Where?  Different slots have the same warning, but this slot needs the warning too.

      is canceled and the object's *finalized* mark might be cleared if
      present.  Currently :term:`CPython` never clears the *finalized* mark;
      this might change in a future version.

   *Automatic finalization* refers to any finalization performed by Python
   except via calls to :c:func:`PyObject_CallFinalizer` or
   :c:func:`PyObject_CallFinalizerFromDealloc`.  No guarantees are made about
   when, if, or how often an object is automatically finalized, except:

   * Python will not automatically finalize an object if it is reachable, i.e.,
     there is a reference to it and it is not a member of a :term:`cyclic
     isolate`.
   * Python will not automatically finalize an object if finalizing it would
     not mark the object as *finalized*.  In :term:`CPython` this currently
     applies to objects whose type does not support garbage collection, i.e.,
     the :c:macro:`Py_TPFLAGS_HAVE_GC` flag is not set.  Such objects can still
     be manually finalized by calling :c:func:`PyObject_CallFinalizer` or
     :c:func:`PyObject_CallFinalizerFromDealloc`.
   * Python will not automatically finalize any two members of a :term:`cyclic

>      isolate` concurrently.

Reviewer 2: I think we're over-documenting for a type slot. Instead of repeating the details of the finalization process, we should just link to the other page with the rest of the garbage collector information.

Author: That was my intention—just document the relationship between the slots in this page, but leave the specifics to the slot-specific documentation.  I added a paragraph at the top to clarify.

.. c:member:: destructor PyTypeObject.tp_finalize

   An optional pointer to an instance finalization function.  This is the C
   implementation of the :meth:`~object.__del__` special method.  Its signature
   is::

      void tp_finalize(PyObject *self);

   The primary purpose of finalization is to perform any non-trivial cleanup
   that must be performed before the object is destroyed, while the object and
   any other objects it directly or indirectly references are still in a

>    consistent state.  See :pep:`442`.  The finalizer is allowed to execute

Reviewer: No, don't look at the original change proposal! It's in the See Also section, that's enough.
```suggestion
   consistent state.  The finalizer is allowed to execute
```

   arbitrary Python code.

   Before Python automatically finalizes an object, some of the object's direct
   or indirect referents might have themselves been automatically finalized.
   However, none of the referents will have been automatically cleared
   (:c:member:`~PyTypeObject.tp_clear`) yet.

   Other non-finalized objects might still be using a finalized object, so the
   finalizer must leave the object in a sane state (e.g., invariants are still
   met).

>    .. warning::

Reviewer: ```suggestion
   .. note::
```

      After Python automatically finalizes an object, Python might start
      automatically clearing (:c:member:`~PyTypeObject.tp_clear`) the object
      and its referents (direct and indirect).  Cleared objects are not
      guaranteed to be in a consistent state; a finalized object must be able
      to tolerate cleared referents.

>    .. warning::

Reviewer: ```suggestion
   .. note::
```

      An object is not guaranteed to be automatically finalized before its
      destructor (:c:member:`~PyTypeObject.tp_dealloc`) is called.  It is
      recommended to call :c:func:`PyObject_CallFinalizerFromDealloc` at the
      beginning of :c:member:`!tp_dealloc` to guarantee that the object is
      always finalized before destruction.

>    .. warning::

Reviewer: ```suggestion
   .. note::
```

      The :c:member:`~PyTypeObject.tp_finalize` function can be called from any
      thread, although the :term:`GIL` will be held.

>    .. warning::

Reviewer: ```suggestion
   .. note::
```
